File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let tempa=[];
    let final=[];
    function backtrack(tempa,nums){
        if(nums.length===0){
            final.push(tempa.slice());
            return;
        }
        for(let i=0;i<nums.length;i++){
            tempa.push(nums[i]);
            nums.splice(i,1);
            backtrack(tempa,nums,final);
            nums.splice(i,0,tempa.pop());

        }
    }
    backtrack(tempa,nums);
    return final;
};

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> permute(int[] nums) {
        boolean[] used = new boolean[nums.length];
        backtrack(new ArrayList<>(), nums, used);
        return res;
    }
    
    private void backtrack(List<Integer> cur, int[] nums, boolean[] used){
        if(cur.size() == nums.length){
            res.add(new ArrayList<>(cur));
            return;
        }
        
        for(int i = 0; i < nums.length; i++){
            if(used[i] == true) continue;
            cur.add(nums[i]);
            used[i] = true;
            backtrack(cur, nums, used);
            cur.remove(cur.size() - 1);
            used[i] = false;
        }
    }
}